prov/lnx/src/lnx_init.c

```diff
@@ -251,300 +264,310 @@ static int lnx_cache_info(struct dlist_entry *head,
+		if (!link_name)
+			return -FI_ENOMEM;
+
+		prov_name = calloc(sizeof(char), 4);
+		if (!prov_name) {
+			free(link_name);
+			return -FI_ENOMEM;
+		}
+		strcpy(prov_name, "lnx");
+		tmp = link_name;
+		tmp2 = link_name + str_len - 1;
+
+		reset = true;
+
+		dlist_foreach_container(&link->link_providers,
+					struct lnx_fi_cache_entry, e, entry) {
+			if (!link->fi_link ||
+			    (!strcmp(link->fi_link->fabric_attr->prov_name,
+				     "shm"))) {
+				if (link->fi_link)
+					fi_freeinfo(link->fi_link);
+				link->fi_link = fi_dupinfo(e->fi);
+				link->fi_link->next = NULL;
+			}
+
+			mr_mode = lnx_update_mr_mode(mr_mode,
+					e->fi->domain_attr->mr_mode, reset);
+			if (reset) {
+				caps = e->fi->caps;
+				min_of_max_msg_size = min_rx_size = min_tx_size = SIZE_MAX;
+				min_iov_limit = LNX_IOV_LIMIT;
+				reset = false;
 			} else {
-				l = fi_copy;
+				caps &= e->fi->caps;
 			}
-			if (current->next->next == head &&
-			    depth == target_depth) {
-				dup = lnx_dupinfo_list(l);
-				if (!dup)
-					return -FI_ENOMEM;
-				new = calloc(1, sizeof(*new));
-				if (!new)
-					return -FI_ENOMEM;
-				new->fi = dup;
-				dlist_init(&new->entry);
-				dlist_insert_tail(&new->entry, result);
+			if (e->fi->ep_attr->max_msg_size < min_of_max_msg_size)
+				min_of_max_msg_size = e->fi->ep_attr->max_msg_size;
+			if (e->fi->tx_attr->iov_limit < min_iov_limit)
+				min_iov_limit = e->fi->tx_attr->iov_limit;
+			if (e->fi->rx_attr->size < min_rx_size)
+				min_rx_size = e->fi->rx_attr->size;
+			if (e->fi->tx_attr->size < min_tx_size)
+				min_tx_size = e->fi->tx_attr->size;
+
+			for (itr = e->fi; itr; itr = itr->next) {
+				len = strlen(itr->domain_attr->name) + 1;
+				if (len < (tmp2 - tmp) - 1) {
+					if (tmp != link_name) {
+						tmp[0] = '+';
+						tmp++;
+					}
+					memcpy(tmp, itr->domain_attr->name, len);
+					tmp += (len - 1);
+				} else {
+					str_len = str_len * 2 + len;
+					tmp = realloc(link_name, str_len);
+					if (!tmp) {
+						free(link_name);
+						return -FI_ENOMEM;
+					}
+					link_name = tmp;
+					tmp += strlen(link_name);
+					tmp2 = link_name + str_len - 1;
+					tmp[0] = '+';
+					tmp++;
+					memcpy(tmp, itr->domain_attr->name, len);
+					tmp += (len - 1);
+				}
 			}
-			rc = gen_links_rec(current->next, head, result, l,
-					   depth+1, target_depth);
-			lnx_remove_tail(&l);
-			if (rc)
-				return rc;
 		}
-		current = current->next;
+
+		link_name = realloc(link_name, strlen(link_name)+1);
+		free(link->fi_link->fabric_attr->prov_name);
+		free(link->fi_link->fabric_attr->name);
+		free(link->fi_link->domain_attr->name);
+		link->fi_link->fabric_attr->prov_name = prov_name;
+		link->fi_link->fabric_attr->name = strdup(prov_name);
+		link->fi_link->domain_attr->name = link_name;
+
+		/* there should only be one fi per link that gets
+		 * propagated up to the caller. This represents the LNX
+		 * link
+		 *
+		 * TODO: What should we do with inject_size?
```

Comment:
Inject size should be the minimum of all providers. You can probably add that to the update info section. If the concern here is shm with FI_HMEM, we can update shm to have the correct inject, not to worry

Reply:
done